### ISSUE TITLE
iframe onload called after every render

### ISSUE THREAD

**Issue by spiffytech** [bikeshaving/crank#181]
I'd like to render an `<iframe>` with a loading indicator, which StackOverflow says should be done by watching for the `load` event on the `iframe` to remove the indicator. When I `refresh()` my component to remove the loading indicator, the `iframe`'s `onload` fires again, triggering another refresh, endlessly. This causes the iframe to infinitely reload the `src` page, hammering the `iframe`'d web server.

I've worked around this by using a variable to track if I need to break the refresh loop (which still loads the `iframe` content twice), but is there a reason the `iframe` fires the `onload` endlessly?

```tsx
function* Iframe(
  this: Context,
) {
  let loading = true;

  for (let props of this) {
    console.log("yielding iframe");
    yield (
      <section>
        {loading && <div>Loading...</div>}
        <iframe
          src="https://httpbin.org/html"
          onload={() => {loading = false; this.refresh()}}
        />
      </section>
    );
  }
}
```

**Comment by spiffytech** [comment]
It looks like I can stop the double render by tracking the last `src` value, comparing it against the one from `props`, and using `<Copy />` and `crank-key` so the `iframe` doesn't get rerendered. Is this expected to be necessary?

**Comment by brainkim** [comment]
>  Is this expected to be necessary?

Nope. It’s definitely an issue, and users shouldn’t be expected to use copies here. The problem is that Crank currently naively sets every property and attribute for each DOM node on render, and `src` for iframes seems to be one of those props which definitely should not be set every render if it hasn’t changed. I was hesitant to change this behavior because at some point I thought that reading and writing to the DOM frequently would be a performance issue. As far as I can tell, the effects of reading props and attributes before writing to them seems to be marginal, and mostly better for performance, because the DOM isn’t as smart as I thought it would be and setting properties like `contentEditable` with the same value will actually invalidate layout, for instance.

I have a working branch for performance which reads the DOM before attempting to set properties and attributes. I’ll try to timebox the work and get it into a 0.3.8 release by tomorrow. 

If you’re wondering why we don’t check the previous props instead of reading the DOM directly, it has to do with one of Crank’s philosophies, which is that we actually treat the DOM as the source of truth rather than the virtual tree. This means we can avoid having to define things like `defaultValue` for uncontrolled components, and it also makes Crank a better fit for a rich text editing backend. I can go more into that if you’re interested.

Sorry for the bug!

**Comment by brainkim** [comment]
Fixed in 0.3.8.

https://codesandbox.io/s/hopeful-grothendieck-i3eww?file=/src/index.js
```jsx
/** @jsx createElement */
import { createElement } from "@bikeshaving/crank";
import { renderer } from "@bikeshaving/crank/dom";

function* IFrame() {
  let loading = true;
  const onload = (ev) => {
    console.log("onload", loading, ev.target.src);
    loading = false;
    this.refresh();
  };

  for (const props of this) {
    console.log("yielding", loading);
    yield (
      <section>
        {loading && <div>Loading...</div>}
        <iframe
          src="https://httpbin.org/html"
          onload={onload}
        />
        <div>
          <button onclick={() => this.refresh()}>
            Refresh
          </button>
        </div>
      </section>
    );
  }
}

renderer.render(<IFrame />, document.body);
```
The only issue now is that the loading event will fire twice in webkit-based browsers (maybe because the onload callback is attached before the DOM node is inserted into the document, and I’m not sure if the first onload call is early because the iframe is loading too quickly. You’ll probably also have to handle src transitions, like maybe you want to see the loading transition again?

The iframe lifecycle is complex enough that it might warrant using manual DOM mutations. Crank provides a `Raw` element which allows you to do this while preserving the component model:
```jsx
/** @jsx createElement */
import { createElement, Raw } from "@bikeshaving/crank";
import { renderer } from "@bikeshaving/crank/dom";

function* IFrame(props) {
  let loading = true;
  const iframe = document.createElement("iframe");
  iframe.src = props.src;
  // manual control over when the onload callback is assigned to the iframe
  iframe.onload = (ev) => {
    console.log("onload", loading, ev.target.src);
    loading = false;
    this.refresh();
  };
  // other imperative initialization logic
  for (props of this) {
    // update logic
    yield (
      <section>
        {loading && <div>Loading...</div>}
        <Raw value={iframe} />
      </section>
    );
  }
}

renderer.render(<IFrame />, document.body);
```

One problem you might run into is that there still isn’t a way to reliably query for when an element has been completely inserted into the DOM. I’m still wrestling with when `schedule()`, `cleanup()`, `ref` callbacks fire, and if we should include a separate context method like `finally()` or `complete()` which fires when DOM nodes are in the document.

Thanks for the report!

**Comment by brainkim** [comment]
This seems to have regressed sometime in 0.4. As a temporary workaround, you can use `c-static` (will be `$static` in 0.5) or Copy. Alternatively, you can also set the src property manually with a ref. Will be fixed in 0.5.

**Comment by brainkim** [comment]
Fixed in 0.5 (hopefully).
